presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -598,17 +612,50 @@ else if (redistributeWrites) {
                     !source.getProperties().isCompatibleTablePartitioningWith(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                     !(source.getProperties().isRefinedPartitioningOver(shufflePartitioningScheme.get().getPartitioning(), false, metadata, session) &&
                             canPushdownPartialMerge(source.getNode(), partialMergePushdownStrategy))) {
+                PartitioningScheme exchangePartitioningScheme = shufflePartitioningScheme.get();
+                if (node.getTablePartitioningScheme().isPresent() && isPrestoSparkAssignBucketToPartitionForPartitionedTableWriteEnabled(session)) {
+                    int writerThreadsPerNode = getTaskPartitionedWriterCount(session);
+                    int bucketCount = getBucketCount(node.getTablePartitioningScheme().get().getPartitioning().getHandle());
+                    int[] bucketToPartition = new int[bucketCount];
+                    int bucket = 0;
+                    int partition = 0;
+                    while (bucket < bucketCount) {
+                        for (int i = 0; i < writerThreadsPerNode && bucket < bucketCount; i++) {
+                            bucketToPartition[bucket] = partition;
+                            bucket++;
+                        }
+                        partition++;
+                    }
```

Comment:
`bucketToPartition[bucket] = bucket / writerThreadsPerNode`

Reply:
This will create a skew. Buckets assigned must be non co-divisible. 

Let me give you an example. Let's say we have 4 buckets and 2 writer threads. That means that we will end up with 2 partitions. Let's assume the buckets are assigned using the `bucketToPartition[bucket] = bucket / writerThreadsPerNode`. `Partition 0` will get `Bucket 0` and `Bucket 2`, `Partition 1` will get `Bucket 1` and `Bucket 3`. Then locally we also take modulo when assigning buckets to threads. `Bucket 0 % 2  = Thread 0`, `Bucket 2 % 2 = Thread 0`. So we will end up with a single thread writing to 2 buckets. What we want to achieve is each thread has a single bucket to write. So we want to assingn `Bucket 0` and `Bucket 1` to `Partition 0`, and `Bucket 2` and `Bucket 3` to `Partition 1`, so further when assigning buckets to threads each Thread has a bucket to write.

Reply:
Please check again, as it is range assigned not modular based distribution

Reply:
Discussed offline. I'm being slow. `bucketToPartition[bucket] = bucket % writerThreadsPerNode`, bucketToPartition[bucket] = bucket / writerThreadsPerNode` is exactly what I'm doing with extra cycles. Let me fix that.